### DISCUSSION TITLE
Possibility to modify nova login form

### DISCUSSION THREAD

**Discussion by mdeimert** [laravel/nova-issues#3954] [Q&A] [answered]
Would be great to have an option to modify the login form for Nova 4.

If not modifying the view we could have an option to modify the e-mail field to something else like an username.

Example use case:

Using LDAP username to login to Nova panel.

**Comment by kichetof** [comment]
Could be great improvement.

Another use case:
Laravel Socialite integration as a login redirector

**Comment by RhysLees** [comment, marked as answer]
This could be a fix for your issue:

You can pass custom auth paths into Nova 4
Awesome if you are using Jetstream and want to leverage 2FA

![image](https://user-images.githubusercontent.com/43909932/163473265-2214990c-5b49-4b5e-985c-4def81dd8b99.png)

**Reply by kichetof** [reply]
Exactly what I use to perform login/logout but I need to serve my own login screen (not a big issue alright, with Laravel Socialite, a simple link is good enough)

**Reply by kobeyy** [reply]
**Note:**
Nova will only use these custom routes when you don't have `withAuthenticationRoutes()` and `withPasswordResetRoutes()` routes registered. 
Your `NovaServiceProvider.php` should have following routes configured.
```
    protected function routes()
    {
        Nova::routes()
                ->register();
    }
```

**Reply by juse-less** [reply]
@kobeyy I think it might be something else causing it when you're changing the lines.
We're doing it in production, and we're able to login, etc, using Jetstream (including 2FA) with the below exact code.
I haven't had issues with it since pre-release of Nova 4 and using Laravel 9.

I ran `artisan route:list`, and, at least in our case, it'll still register the Nova routes, but so will Fortify (as part of Jetstream).
We do however use `/login` and not Nova's path-prefix, if that might be the issue you're facing?
If I do visit `/nova/login` (or `/admin/login` in our case), it'll redirect to `/login`, though.

```php
protected function routes(): void
{
    Nova::routes()->withAuthenticationRoutes()->register();
}
```

**Reply by kobeyy** [reply]
@juse-less interesting, maybe its because Jetstream overwrites part of the nova javascript. Have not yet worked with Jetstream myself.
I based my findings on part of the nova source code. For example the logout function (`vendor/laravel/nova/resources/js/store/nova.js`) looks like this.

```javascript
async logout({ state }, customLogoutPath) {
    let response = null

    if (!Nova.config('withAuthentication') && customLogoutPath) {
      response = await Nova.request().post(customLogoutPath)
    } else {
      response = await Nova.request().post(Nova.url('/logout'))
    }

    return response?.data?.redirect || null
}
```

**Reply by juse-less** [reply]
@kobeyy I see. That is a bit interesting!
Depending on if `Nova.url()` prefixes with Nova's path-prefix or not, it sounds like Nova calls /logout in our case, which is registered by Jetstream.
Then that also sounds like I can technically remove the routes config we have.

I can try it later in the evening, hopefully.
In the meantime, this is the result of our [login](https://app.warp.dev/block/nD3rUu1p4AQC2wbJFm6WFL) and [logout](https://app.warp.dev/block/5mKdgzz7wV2gfqWKGkTF36) routes (we don't do password resets, etc).
And our routes Nova config (inside config/nova.php)
```php
'routes' => [
    'login' => '/login',
    'logout' => '/logout',
],
```

Most of our Nova config is default, including domain and guard.

**Reply by juse-less** [reply]
Disclaimer: I didn't read any code or follow its logic at this time. I merely quickly enabled/disabled config and route registration, as well as confirmed output with `artisan route:list`, and visited the authentication page for visual 'testing'.
So this is just my quick experience and impression.

---
I ran `artisan optimize:clear` between all steps below, to ensure all caches and routes were always clean.
I also closed the browser and opened a new Incognito window between each step, to ensure the browser wasn't playing tricks on me.

1. Using the below code instead, does remove the [login](https://app.warp.dev/block/hnJS9N6HMugBKTIImOqDxL) and [logout](https://app.warp.dev/block/igiYPfJbFtOlwfS0NDLccI) routes for Nova (which is expected).
    And I'm still redirected and everything as before.
    ```php
    protected function routes(): void
    {
        Nova::routes()->register();
    }
    ```
1. Removing the `routes` section in the Nova config, does give me a 404 on /admin/login, which should also be expected.
1. Re-adding the route registration in Nova does re-register the routes, as expected.
    I now have both Fortify and Nova [login](https://app.warp.dev/block/vTkQafixYStXJB4TWhpk0n) and [logout](https://app.warp.dev/block/sRDSEzbPAPaQWqMTPjZGrU) routes again.
    Visiting /admin (in my case) will give me the Nova authentication page, instead of Fortify's/Jetstream's.
    ```php
    protected function routes(): void
    {
        Nova::routes()->withAuthenticationRoutes()->register();
    }
    ```
1. Re-adding the `routes` section in the Nova config, redirects me to my normal authentication page (Jetstream's) when visiting /admin and even /admin/login.

So, to summarise on my experience/impression, @kobeyy is right on the code he read (and posted above) - you need to omit registering the Nova routes to use your custom routes authentication _logic_.
However, in doing so, you get the Nova authentication page.
By using _both_ custom routes [logic], as well as registering Nova's routes, Nova will actually redirect you to your custom routes' authentication page(s), which will then also use the authentication logic for that page.
Nova's authentication page routes are still registered, but I can't seem to access the actual pages as I'm being redirected to my custom ones.
Even when manually writing the whole URI, as seen in `artisan route:list --name=login`, for instance.

So the difference here seems to be authentication _logic_ vs authentication _pages_.

I didn't try using Jetstream's 2FA through Nova's authentication page, as I have it disabled locally.

---
Edit: It's actually better to use `artisan route:list --path=login`, because Fortify doesn't name their authentication logic route, leaving it out of the linked results above.
It's a POST route, just like Nova's own.
```
> artisan route:list --path=login
...
POST       admin/login .......................... nova.login › Laravel\Nova › LoginController@login
POST       login ................................ Laravel\Fortify › AuthenticatedSessionController@store
...
```

**Reply by chrillep** [reply]
So if you dont have/want Jetstream. What to do ? 😄 
from https://jetstream.laravel.com/3.x/installation.html#installing-jetstream
<img width="765" alt="Skärmavbild 2023-03-25 kl  16 59 34" src="https://user-images.githubusercontent.com/1267931/227728321-d0f648be-fa46-495b-820b-d28e3b0afcb6.png">


How do we modify the view of the login page? As that was the OP question.

**Reply by RhysLees** [reply]
> So if you dont have/want Jetstream. What to do ? 😄 from https://jetstream.laravel.com/3.x/installation.html#installing-jetstream <img alt="Skärmavbild 2023-03-25 kl 16 59 34" width="765" src="https://user-images.githubusercontent.com/1267931/227728321-d0f648be-fa46-495b-820b-d28e3b0afcb6.png">
> 
> How do we modify the view of the login page? As that was the OP question.

You dont have to use jetstream, you can make your own authentication logic/pages with fortify or any way you want and use the same config and it will work

**Reply by chrillep** [reply]
I'm currently not using nova, but you still have to maintain/make your own login page ?
OP Question is: modify nova default one 😄 .
The answer is you can't? Should not?

Im not trying to be brash. I just want it to be contextually correct. ❤️ 
Other solutions work but its not what was asked for.

**Comment by hussainanjar** [comment]
A possible implementation of socialite would really help

**Reply by mihaileu** [reply]
I really need it

**Comment by chrillep** [comment]
what we want is this tutorial to work in nova 4.
https://medium.com/@SlyFireFox/laravel-nova-implementing-social-logins-41ea36406f6c
https://github.com/peterfox/nova-social-login-demo
https://github.com/peterfox/nova-social-login-demo/issues/1

i guess you need to write your own tool that overrides the Login component `vendor/laravel/nova/resources/js/pages/Login.vue`
as shown in https://github.com/laravel/nova-issues/issues/5119#issuecomment-1328150313

would be nice if we could pass data/components to the Default Nova Login Component via Slot or similar instead.
Then you could just pass a list of buttons for every service.

**Comment by scramatte** [comment]
Hi,

Currently, we use the following package and it works great. We can change login page and including create new pages for example to sign up

**https://github.com/Muetze42/nova-assets-changer**

Our app use LDAP authentication using LdapRecord
**https://ldaprecord.com/docs/laravel/v2**

The only problem is during migration as "getKeyType" method is called on the auth provider.  If you try to use anything different from standard eloquent you will raise an exception.

Please see the following issue for details. The problem is due to something that is hard-coded in nova ...
https://githubcom/DirectoryTree/LdapRecord-Laravel/issues/423

Anyway, it will better to have such features natively without requires any add-ons.

Regards

**Reply by chrillep** [reply]
Sounds good! could you add a code example using https://github.com/Muetze42/nova-assets-changer to swap Login Component `vendor/laravel/nova/resources/js/pages/Login.vue` ?

ref: 

- https://github.com/Muetze42/nova-assets-changer/issues/2

code example

- https://github.com/Muetze42/nova-assets-changer/blob/main/docs/register-page.md
- https://github.com/Muetze42/nova-assets-changer-register-page
